### PR TITLE
Graceful NATS Shutdown [go]

### PR DESCRIPTION
**Pull request by natewoods-wf**
### Story:
We've noticed that when a frugal server stops, it drops all in-memory messages leading to dropped requests. This is less than ideal in communication heavy environments.  We'd like to remove the nats topic interest but continue processing any received requests.

This PR makes the following changes:

1. `Stop` now waits for the nats subscription to remove interest and complete queuing in-memory requests
2. `Serve` now waits for all workers to complete processing

### Acceptance Criteria:
- [x] Code has been tested and results documented
- [x] Unit tests have been updated
- [x] Updates to documentation if necessary
- [x] Verify and document changes to any other Messaging components
- [x] Pull request made against the 'develop' branch, not master

### Design Notes:
1. The Nats library uses an in-memory queue of messages that must be queued into frugal's `workC` before existing `Serve` (and `Stop`).  There are two methods to do this:
   * Spin until each subscription returns `false` for `IsValid`
      * Spin-loops in code aren't exactly awesome
   * Invoke a "barrier" to the full nats connection, requiring all subscriptions to acknowledge the barrier
      * Less loops in code, and it doesn't affect any processing go-routines

### How To Test:
Included unit test demonstrates the failure before and after these changes.  Similarily, it demonstrates the necessity of the "barrier".

### My Test Results:

<details>
<summary>Before Changes</summary>

```
$ go test -v -timeout 10s -run Shutdown -count=1
=== RUN   TestShutdown
INFO[0000] frugal: server running...                    
INFO[0000] frugal: server stopping...                   
    shutdown_test.go:78: Received 10; Processed: 0; Remaining: 2
    shutdown_test.go:79: 
        	Error Trace:	/home/natewoods/wk/messaging-shutdown/pkg/frugal/shutdown_test.go:79
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 10
        	Test:       	TestShutdown
--- FAIL: TestShutdown (0.01s)
FAIL
exit status 1
FAIL	github.com/Workiva/frugal/lib/go	0.019s
```
</details>

<details>
<summary>Without Barrier</summary>

```
$ go test -v -timeout 10s -run Shutdown -count=1
=== RUN   TestShutdown
INFO[0000] frugal: server running...                    
INFO[0000] frugal: server stopping...                   
DEBU[0000] frugal: subscriptions removed                
DEBU[0000] frugal: connection flushed                   
panic: send on closed channel

goroutine 9 [running]:
github.com/Workiva/frugal/lib/go.(*fNatsServer).handler(0xc000214180, 0xc0003c81c0)
	/home/natewoods/wk/messaging-shutdown/pkg/frugal/nats_server.go:300 +0xdd
github.com/nats-io/nats%2ego.(*Conn).waitForMsgs(0xc0002b4700, 0xc000216a90)
	/home/natewoods/go/pkg/mod/github.com/nats-io/nats.go@v1.26.0/nats.go:3001 +0x442
created by github.com/nats-io/nats%2ego.(*Conn).subscribeLocked
	/home/natewoods/go/pkg/mod/github.com/nats-io/nats.go@v1.26.0/nats.go:4232 +0x3f8
exit status 2
FAIL	github.com/Workiva/frugal/lib/go	0.020s
```
</details>

<details>
<summary>After Changes</summary>

```
$ go test -v -timeout 10s -run Shutdown -count=1
=== RUN   TestShutdown
INFO[0000] frugal: server running...                    
INFO[0000] frugal: server stopping...                   
DEBU[0000] frugal: subscriptions removed                
DEBU[0000] frugal: connection flushed                   
DEBU[0000] frugal: subscription queues drained           took=301.916813ms
DEBU[0000] frugal: workers completed                    
    shutdown_test.go:78: Received 10; Processed: 10; Remaining: 0
--- PASS: TestShutdown (0.52s)
```
</details>

#### Reviewers:
@Workiva/service-platform

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Review by brettkail-wk — APPROVED** [review]
My knowledge of Go (and this area of Frugal Go is weak), but the high-level looks ok.

File: lib/go/nats_server.go [hunk 1]
```diff
@@ -212,25 +214,76 @@ func (f *fNatsServer) Serve() error {
...
+	for _, sub := range subs {
+		if err := sub.Drain(); err != nil {
+			return err
+		}
 	}
+	log.Debug(`frugal: subscriptions removed`)
 
+	// push subscription removals to the nats server
+	// technically, sub.Drain called `kickFlusher`, but this explicitly waits for the PONG
+	if err := f.conn.Flush(); err != nil {
```
[the remark this answers is not in the document]
Comment (natewoods-wf) [reply lib/go/nats_server.go:260]: Hey @brettkail-wk, I only found this after digging through [the nats.go library](https://github.com/nats-io/nats.go) and watching debug traces of the nats server.  Not sure if the [nata.java library](https://github.com/nats-io/nats.java) has the same shortcoming.  Might need some digging into.

For reference, [Subscription.Drain] calls [Conn.unsubscribe] calls [Conn.kickFlusher].

[Subscription.Drain]: https://github.com/nats-io/nats.go/blob/21df434ace1dd2a597d478aff0589cf5e060c1a3/nats.go#L4335-L4345
[Conn.unsubscribe]: https://github.com/nats-io/nats.go/blob/21df434ace1dd2a597d478aff0589cf5e060c1a3/nats.go#L4455-L4494
[Conn.kickFlusher]: https://github.com/nats-io/nats.go/blob/21df434ace1dd2a597d478aff0589cf5e060c1a3/nats.go#L3537-L3546

[the remark this answers is not in the document]
Comment (brettkail-wk) [reply lib/go/nats_server.go:260]: I believe all Frugal NATS clients have that issue for subscribe/unsubscribe.  I remember hitting issues with a service that published a NATS message in response to an HTTP POST, so rapidly subscribe+POST would miss a message.

**Review by brettkail-wk — APPROVED** [review]

File: lib/go/nats_server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (natewoods-wf) [reply lib/go/nats_server.go:263]: Hey @brettkail-wk, I dug through the code on the java side as well.  Looks like java [only enqueues the subscription or un-subscription into a write buffer](https://github.com/nats-io/nats.java/blob/3ab4cd483add761305253ece544d8123dd2ed658/src/main/java/io/nats/client/impl/NatsConnection.java#L1543).  So a similar problem would show in java as well.  I don't have a ton of experience in java, so I've added a few comments to the code in hopes that someone in the future can track them down and correct them.  Thanks for the feedback!

**Review by adamdockter-wk — APPROVED** [review]

**Comment by natewoods-wf** [comment]
@Workiva/release-management-p for merge

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**